backend/src/baserow/core/formula/field.py

```diff
@@ -1,11 +1,488 @@
-from django.db import models
+import json
+import logging
+from typing import Dict, List, Union
+
+from django.db import connection, models
+
+from baserow.core.formula import BaserowFormulaObject
+from baserow.core.formula.types import (
+    BASEROW_FORMULA_MODE_SIMPLE,
+    BaserowFormulaMinified,
+    FormulaFieldDatabaseValue,
+    JSONFormulaFieldDatabaseValue,
+    JSONFormulaFieldResult,
+)
+
+logger = logging.getLogger(__name__)
+
+
+BASEROW_FORMULA_VERSION_INITIAL = "0.1"
 
 
 class FormulaField(models.TextField):
     """
-    A formula field contains the text value of a runtime formula like:
-    - concat("test:", get("page_parameter.id"), "-")
-    - get("data_source.Product.id")
+    A formula field which can contain:
 
-    For now it's just a text field but we can add layer of validation later.
+    - A JSON-serialized formula string:
+        - E.g. 'get(\"data_source.123.field_123\")'
+        - This can happen if a user fetches a row with one or more formula fields
+          which were not yet migrated to the new formula context format.
+    - A JSON-serialized formula context:
+        - E.g. {"f":"get(\"data_source.123.field_123\")\","m": "simple","v":"0.1"}
+        - This is the new format which contains the formula string, the mode (simple,
+            advanced, raw) and the version of the formula context.
     """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # For compat reasons, applied the parameters we used to receive.
+        # These can be altered once we inherit from `JSONField`.
+        self.default = ""
+        self.null = True
+        self.blank = True
+
+    def _value_is_serialized_object(self, value: FormulaFieldDatabaseValue) -> bool:
+        return isinstance(value, str) and value[:1] == "{" and value[-1:] == "}"
```

Comment:
[non-blocking opinion] I don't feel strongly, but I think this form is a bit clearer:
```suggestion
        return isinstance(value, str) and value.startswith("{") and value.endswith("}")
```